### PR TITLE
0.4.19 release notes

### PR DESCRIPTION
**Pull request by Czaki**
Initial patches for 0.4.19 release

### CODE REVIEW

File: highlight/0.4.19.md
Comment (Czaki) [inline highlight/0.4.19.md:1]: @melonora I'm not proud from this text, I think that we should treat its as placeholder.

File: highlight/0.4.19.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+as not all the features are ready yet ([napari/napari/#6283](https://github.com/napari/napari/pull/6283)).
+
+Secondly we make napari pydantic 2 compatible.
+At this moment we are using `pydantic.v1` to achieve this.
+In this release the bundle will be shipped with pydantic 1, but
+we expect that in the next release we will ship bundle with pydantic 2.
+So if your plugin is using pydantic please make sure that it is compatible with pydantic 2.
+([napari/napari/#6358](https://github.com/napari/napari/pull/6358))
+
+Lastly we have fixed performance problem with label layer by move part of calculation on GPU
```
Comment (melonora) [inline highlight/0.4.19.md:13]:
```suggestion
Lastly, we have fixed a performance problem of the label layer by moving part of calculations to the GPU
```

File: highlight/0.4.19.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+([napari/napari/#3308](https://github.com/napari/napari/pull/3308))
```
Comment (melonora) [inline highlight/0.4.19.md:14]:
```suggestion
This release is mostly a bug fix release and hopefully the latest before all the excited features that are coming with version 0.5.0.

First of all we have an important announcement with this release.
We have decided to postpone the `viewer.window.qt_viewer` to 0.6.0 release
as not all the features are ready yet ([napari/napari/#6283](https://github.com/napari/napari/pull/6283)).

Furthermore we would like to highlight the following changes:

 We have made napari pydantic 2 compatible.
At this moment we are using `pydantic.v1` to achieve this.
In this release the bundle will be shipped with pydantic 1, but
we expect that in the next release we will ship the bundle with pydantic 2.
Please, if you are a plugin developer and your plugin uses pydantic, ensure 
that it is compatible with pydantic 2 ([napari/napari/#6358](https://github.com/napari/napari/pull/6358)).

Also, we have fixed a performance problem of the label layer by moving part of calculations to the GPU
([napari/napari/#3308](https://github.com/napari/napari/pull/3308)).

If you have any questions or suggestions regarding napari core, for example how to adapt your plugin to be compatible with future napari versions, please get in touch! We have weekly community meetings, for which you can find the schedule [here](https://napari.org/stable/community/meeting_schedule.html). We would like to hear from you.
```

File: highlight/0.4.19.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+At this moment we are using `pydantic.v1` to achieve this.
+In this release the bundle will be shipped with pydantic v1, but
+we expect that in the next release we will ship the bundle with pydantic v2.
+Please, if you are a plugin developer and your plugin uses pydantic, ensure 
+that it is compatible with pydantic v2 ([napari/napari/#6358](https://github.com/napari/napari/pull/6358)).
+
+Also, we have fixed a performance problem of the label layer by moving part of calculations to the GPU
+([napari/napari/#3308](https://github.com/napari/napari/pull/3308)).
+
+If you have any questions or suggestions regarding napari core, for example how to adapt your plugin to be compatible with future napari versions, please get in touch! We have weekly community meetings, for which you can find the schedule [here](https://napari.org/stable/community/meeting_schedule.html). We would like to hear from you.
```
Comment (brisvag) [inline highlight/0.4.19.md:19]:
```suggestion
This release is mostly a bug fix release. We aim for the next release to be 0.5.0, with many new exciting features!

Highlights in `v0.4.19`:

The removal of public access to `viewer.window.qt_viewer` was postponed to `v0.6.0` ([napari/napari/#6283](https://github.com/napari/napari/pull/6283)).

Napari is now pydantic 2 compatible. At this moment we are using `pydantic.v1` to achieve this. In this release the bundle will be shipped with pydantic v1, but
we expect that in the next release we will ship the bundle with pydantic v2.
Please, if you are a plugin developer and your plugin uses pydantic, ensure 
that it is compatible with pydantic v2 ([napari/napari/#6358](https://github.com/napari/napari/pull/6358)).

Performance of the `Labels` layer is drastically improved for many labels by moving part of calculations to the GPU ([napari/napari/#3308](https://github.com/napari/napari/pull/3308)).

If you have any questions or suggestions regarding napari core, for example on how to adapt your plugin to be compatible with future napari versions, please get in touch! We have weekly community meetings, for which you can find the schedule [here](https://napari.org/stable/community/meeting_schedule.html). We would like to hear from you.
```

Comment (Czaki) [reply highlight/0.4.19.md:19]: Do you prefer to not mention the reason why qt_viewer deprecation was postponed?

Comment (brisvag) [reply highlight/0.4.19.md:19]: I don't think it's particularly useful in the highlights, but I'm totally fine to leave it in if you prefer!

Comment (Czaki) [reply highlight/0.4.19.md:19]: I think that it may be a good suggestion to tell plugin maintainers that they could ask for some API that they require. But I may be wrong.

Comment (brisvag) [reply highlight/0.4.19.md:19]: Good idea!

Comment (Czaki) [reply highlight/0.4.19.md:19]: "private or deprecated api" ?

File: highlight/0.4.19.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
+This release is mostly a bug fix release. We aim for the next release to be 0.5.0, with many new exciting features!
+
+Highlights in `v0.4.19`:
+
+The removal of public access to `viewer.window.qt_viewer` was postponed to `v0.6.0` ([napari/napari/#6283](https://github.com/napari/napari/pull/6283)). If you're interested in some feature that's now hidden behind private or deprecated API, please make an issue on the napari repo requesting the feature to be exposed!
```
Comment (psobolewskiPhD) [inline highlight/0.4.19.md:5]:
```suggestion
The removal of public access to `viewer.window.qt_viewer` has been postponed to `v0.6.0` ([napari/napari/#6283](https://github.com/napari/napari/pull/6283)). If you're interested in some feature that's currently behind private or deprecated API, please make an issue on the napari repo requesting the feature to be exposed!
```

File: highlight/0.4.19.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+Napari is now pydantic 2 compatible. At this moment we are using `pydantic.v1` to achieve this. In this release the bundle will be shipped with pydantic v1, but
```
Comment (psobolewskiPhD) [inline highlight/0.4.19.md:7]:
```suggestion
napari is now pydantic 2 compatible. At this moment we are using `pydantic.v1` to achieve this. Additionally, for this release, the napari conda bundle will be shipped with pydantic v1, but
```

**Comment by psobolewskiPhD** [comment]
Small copy edits.

File: highlight/0.4.19.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+At first we have decided to postpone the `viewer.window.qt_viewer` to 0.6.0 release
+as not all the features are ready yet ([napari/napari/#6283](https://github.com/napari/napari/pull/6283)).
+
+Secondly we make napari pydantic 2 compatible.
+At this moment we are using `pydantic.v1` to achieve this.
+In this release the bundle will be shipped with pydantic 1, but
+we expect that in the next release we will ship bundle with pydantic 2.
+So if your plugin is using pydantic please make sure that it is compatible with pydantic 2.
+([napari/napari/#6358](https://github.com/napari/napari/pull/6358))
```
Comment (Carreau) [inline highlight/0.4.19.md]:
```suggestion
we expect the next release to ship with pydantic 2 bundled.
If your plugin is using pydantic please make sure that it is compatible with pydantic 2.
([napari/napari/#6358](https://github.com/napari/napari/pull/6358))
```
?

File: highlight/0.4.19.md (hunk 1 above)
Comment (Carreau) [inline highlight/0.4.19.md]:
```suggestion
Lastly we have fixed performance problem with label layer by moving part of calculation on GPU
```

File: highlight/0.4.19.md (hunk 6 above)
Comment (Czaki) [reply highlight/0.4.19.md]: nope. we will not bundle pydantic 2 in napair. We will ship pydantic 2 in a bundle installer (so it will affect only downloadable executables, not installed from pip/conda).

File: highlight/0.4.19.md (hunk 1 above)
Comment (Czaki) [reply highlight/0.4.19.md]: Hm. I think that we will need to rewrite this part totally as we now working on updating of mechanism.

**Comment by brisvag** [comment]
This seems like way too much details for the highlights. I would say just enough so people know to use smaller unsigned dtypes if performance is an issue, but no need to explain the how/why and point to all the PR in the highlights.

**Comment by psobolewskiPhD** [comment]
Is it worth reminding people of the `Convert dtype` contextual menu? or is that too far afield.

**Comment by Czaki** [comment]
> Is it worth reminding people of the `Convert dtype` contextual menu? or is that too far afield.

I think yes. 



> This seems like way too much details for the highlights. I would say just enough so people know to use smaller unsigned dtypes if performance is an issue, but no need to explain the how/why and point to all the PR in the highlights.

Maybe yes. I'm not best text writter

File: highlight/0.4.19.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
+This release is mostly a bug fix release. We aim for the next release to be 0.5.0, with many new exciting features!
```
Comment (brisvag) [inline highlight/0.4.19.md:1]: Just an idea to rephrase this to make it clear we're already stocked up on features for 0.5.0 :)

```suggestion
This release contains mostly a lot of bug fixes and improvements, as we prepare for the next, feature-rich release 0.5.0!
```

File: highlight/0.4.19.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+Highlights in `v0.4.19`:
+
+A new feature: [napari/napari#6102](https://github.com/napari/napari/pull/6102) added convenient parsing of color names and hex color strings in layer data
+and generating black-to-color colormaps when adding layers to the viewer.
+For example, one can now use: `viewer.add_image(data, colormap="crimson")` or `viewer.add_image(data, colormap="#FAFA00").
+To see what color names can be handled, please refer to the
+[vispy` dictionary of colors](https://github.com/vispy/vispy/blob/269ed1ac4d8126421fd5a7eb06a2996d63f46b17/vispy/color/_color_dict.py#L181).
+Note: there is a namespace collision for the case of `orange` which is both a proper `vispy` colormap name and a color name.
+Previously, `orange` was handled as the `vispy` colormap, which is from white to orange.
+However, with 0.4.19, `orange` will be handled as a color and the generated colormap will be from black to orange.
```
Comment (brisvag) [inline highlight/0.4.19.md:12]:
```suggestion
Colormaps: [napari/napari#6102](https://github.com/napari/napari/pull/6102) added convenient parsing of color names and hex color strings in layer data
and generating black-to-color colormaps when adding layers to the viewer.
For example, one can now use: `viewer.add_image(data, colormap="crimson")` or `viewer.add_image(data, colormap="#FAFA00").
To see what color names can be handled, please refer to the
[vispy` dictionary of colors](https://github.com/vispy/vispy/blob/269ed1ac4d8126421fd5a7eb06a2996d63f46b17/vispy/color/_color_dict.py#L181).
Note: there is a namespace collision for the case of `orange` which is both a proper `vispy` colormap name and a color name.
Previously, `orange` was handled as the `vispy` colormap, which is from white to orange.
However, with 0.4.19, `orange` will be handled as a color and the generated colormap will be from black to orange.
```

File: highlight/0.4.19.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+For bigger data types we need to convert data, so it leads to bigger memory usage.
+Also best performance for bigger data requires to install numba library.
+
+So if you wold like for best user performance, please try to use the minimal possible data type.
+Even if, in the future, we'll use more advanced techniques for rendering labels on GPU, that will allow
+transferring any data without preprocessing, the GPU often offers less VRAM that computer has RAM.
+Also we experiment a little with current options and increase the size of
+send data (uint8, uint16 and float32) decrease performance of rendering on GPU.
+The test was performed on NVIDIA GeForce RTX 3060 with 12GB VRAM.
```
Comment (brisvag) [inline highlight/0.4.19.md:44]: Most of this is too much for the highlights in my opinion. What matters to users is: labels are better, especially if you use the smallest dtypes. I'm not even sure we should put the list of PRs; isnt there a master issue that links to all of them?

```suggestion
We have fixed several issues causing labels to be rendered with incorrect colors. As an added benefit, this also improved the rendering performance of labels! To maximize performance, make sure to use the smallest possible data type that can fit your labels! (e.g: (u)int8, (u)int16). Otherwise, if you still need bigger data types, installing the optional dependency `numba` may help with your performance.
This work took several pull requests: you can follow the process here:
[napari/napari/#3308](https://github.com/napari/napari/pull/3308),
[napari/napari/#6411](https://github.com/napari/napari/pull/6411),
[napari/napari/#6467](https://github.com/napari/napari/pull/6467),
[napari/napari/#6439](https://github.com/napari/napari/pull/6439),
[napari/napari/#6607](https://github.com/napari/napari/pull/6607).
```

**Review by brisvag — APPROVED** [review]

File: highlight/0.4.19.md (hunk 9 above)
Comment (Czaki) [reply highlight/0.4.19.md:44]: There is no master issue, as part of them was created in response to bug pointed by users. 
@jni did you have an opinion?

File: highlight/0.4.19.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+[napari/napari/#3308](https://github.com/napari/napari/pull/3308),
+[napari/napari/#6411](https://github.com/napari/napari/pull/6411),
+[napari/napari/#6467](https://github.com/napari/napari/pull/6467),
+[napari/napari/#6439](https://github.com/napari/napari/pull/6439).
+[napari/napari/#6607](https://github.com/napari/napari/pull/6607).
+This also improves performance of rendering labels.
+However, there are still some limitations.
+Best performance is for labels encoded as (u)int8.
+As GPU API accept 8 and 16 bits int textures and float textures,
+for labels stored as (u)int8 and (u)int16 we could transfer data to GPU without preprocessing.
```
Comment (melonora) [inline highlight/0.4.19.md:34]:
```suggestion
As the GPU API accepts 8 and 16 bits int textures and float textures,
for labels stored as (u)int8 and (u)int16 we could transfer data to GPU without pre-processing.
```

File: highlight/0.4.19.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+For bigger data types we need to convert data, so it leads to bigger memory usage.
+Also best performance for bigger data requires to install numba library.
+
+So if you wold like for best user performance, please try to use the minimal possible data type.
+Even if, in the future, we'll use more advanced techniques for rendering labels on GPU, that will allow
+transferring any data without preprocessing, the GPU often offers less VRAM that computer has RAM.
+Also we experiment a little with current options and increase the size of
+send data (uint8, uint16 and float32) decrease performance of rendering on GPU.
+The test was performed on NVIDIA GeForce RTX 3060 with 12GB VRAM.
```
Comment (melonora) [inline highlight/0.4.19.md:44]:
```suggestion
To achieve optimal user performance, it is advised to utilize the most compact data type possible.
Although future rendering techniques on the GPU may enable the direct transfer of data without pre-processing, it's important to consider that GPUs often have less VRAM compared to the system's RAM. 
Our experiments indicate that increasing the size of the sent data (such as using uint8, uint16, and float32) can negatively impact rendering performance on GPUs. These tests were performed on an NVIDIA GeForce RTX 3060 with 12GB VRAM.
```

File: highlight/0.4.19.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+[napari/napari/#6439](https://github.com/napari/napari/pull/6439).
+[napari/napari/#6607](https://github.com/napari/napari/pull/6607).
+This also improves performance of rendering labels.
+However, there are still some limitations.
+Best performance is for labels encoded as (u)int8.
+As GPU API accept 8 and 16 bits int textures and float textures,
+for labels stored as (u)int8 and (u)int16 we could transfer data to GPU without preprocessing.
+
+For bigger data types we need to convert data, so it leads to bigger memory usage.
+Also best performance for bigger data requires to install numba library.
```
Comment (melonora) [inline highlight/0.4.19.md:37]:
```suggestion
For bigger data types we need to convert data, which leads to bigger memory usage.
For best performance when working with bigger data, it is required to install the numba library.
```

File: highlight/0.4.19.md (hunk 9 above)
Comment (melonora) [reply highlight/0.4.19.md:44]: I largely agree with @brisvag. However, for some plugin developers this could still be useful information no?

Comment (brisvag) [reply highlight/0.4.19.md:44]: Sure, but maybe we should have a separate section for the details, or just make a new issue with this info and the links to the PRs and point to that. This just feels noisy and useless to most people who just need to know "int8 better".

File: additional_notes/0.4.19/highlights.md [hunk 13]
```diff
@@ -0,0 +1,111 @@
+This release mostly contains a lot of bug fixes and performance improvements.
+But look out for 0.5.0, coming to a software repository near you — we expect to
+release a lot of new features then!
+
+### BIG improvements to the Labels layer
+
+[#3308](https://github.com/napari/napari/pull/3308) closed *many* long-standing
+bugs in the handling of colors in the Labels layer: the color swatch in the
```
Comment (brisvag) [inline additional_notes/0.4.19/highlights.md:8]:
```suggestion
bugs in the handling of colors in the Labels layer: the color switch in the
```

**Review by brisvag — APPROVED** [review]
1 typo, but other than that love the new notes!

File: additional_notes/0.4.19/highlights.md (hunk 13 above)
Comment (jni) [reply additional_notes/0.4.19/highlights.md:8]: nope, swatch is correct: https://www.merriam-webster.com/dictionary/swatch

**Comment by jni** [comment]
@Czaki Thanks for the script updates, they look great. Can I push here again? I'll move the API changes note to the relevant section, and I'll make a note in the deprecations section for napari/napari#6542.

**Comment by Czaki** [comment]
yes

**Comment by jni** [comment]
Ok. I put a space in `api changes.md` based on my reading of the code, but obviously, feel free to update as needed. As far as I'm concerned this in now ready! 😍 🚀

File: additional_notes/0.4.19/highlights.md (hunk 13 above)
Comment (brisvag) [reply additional_notes/0.4.19/highlights.md:8]: uh, the more you know!
